commands/apps.go

```diff
@@ -0,0 +1,393 @@
+	}
+
+	appSpec, err := parseAppSpec(specBytes)
+	if err != nil {
+		return err
+	}
+
+	app, err := c.Apps().Update(id, &godo.AppUpdateRequest{Spec: appSpec})
+	if err != nil {
+		return err
+	}
+	notice("App updated")
+
+	return c.Display(displayers.Apps{app})
+}
+
+// RunAppsDelete deletes an app.
+func RunAppsDelete(c *CmdConfig) error {
+	if len(c.Args) < 1 {
+		return doctl.NewMissingArgsErr(c.NS)
+	}
+	id := c.Args[0]
+
+	err := c.Apps().Delete(id)
+	if err != nil {
+		return err
+	}
+	notice("App deleted")
+
+	return nil
+}
+
+// RunAppsCreateDeployment creates a deployment for an app.
+func RunAppsCreateDeployment(c *CmdConfig) error {
+	if len(c.Args) < 1 {
+		return doctl.NewMissingArgsErr(c.NS)
+	}
+	appID := c.Args[0]
+
+	deployment, err := c.Apps().CreateDeployment(appID)
+	if err != nil {
+		return err
+	}
+	notice("Deployment created")
+
+	return c.Display(displayers.Deployments{deployment})
+}
+
+// RunAppsGetDeployment gets a deployment for an app.
+func RunAppsGetDeployment(c *CmdConfig) error {
+	if len(c.Args) < 2 {
+		return doctl.NewMissingArgsErr(c.NS)
+	}
+	appID := c.Args[0]
+	deploymentID := c.Args[1]
+
+	deployment, err := c.Apps().GetDeployment(appID, deploymentID)
+	if err != nil {
+		return err
+	}
+
+	return c.Display(displayers.Deployments{deployment})
+}
+
+// RunAppsListDeployments lists deployments for an app.
+func RunAppsListDeployments(c *CmdConfig) error {
+	if len(c.Args) < 1 {
+		return doctl.NewMissingArgsErr(c.NS)
+	}
+	appID := c.Args[0]
+
+	deployments, err := c.Apps().ListDeployments(appID)
+	if err != nil {
+		return err
+	}
+
+	return c.Display(displayers.Deployments(deployments))
+}
+
+// RunAppsGetLogs gets app logs for a given component.
+func RunAppsGetLogs(c *CmdConfig) error {
+	if len(c.Args) < 3 {
+		return doctl.NewMissingArgsErr(c.NS)
+	}
+	appID := c.Args[0]
+	deploymentID := c.Args[1]
+	component := c.Args[2]
+
+	logTypeStr, err := c.Doit.GetString(c.NS, doctl.ArgAppLogType)
+	if err != nil {
+		return err
+	}
+	var logType godo.AppLogType
+	switch logTypeStr {
+	case strings.ToLower(string(godo.AppLogTypeBuild)):
+		logType = godo.AppLogTypeBuild
+	case strings.ToLower(string(godo.AppLogTypeDeploy)):
+		logType = godo.AppLogTypeDeploy
+	case strings.ToLower(string(godo.AppLogTypeRun)):
+		logType = godo.AppLogTypeRun
+	default:
+		return fmt.Errorf("Invalid log type %s", logTypeStr)
+	}
+
+	logs, err := c.Apps().GetLogs(appID, deploymentID, component, logType)
+	if err != nil {
+		return err
+	}
+
+	if logs.LiveURL != "" {
+		resp, err := http.Get(logs.LiveURL)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		io.Copy(c.Out, resp.Body)
+	} else if len(logs.HistoricURLs) > 0 {
+		resp, err := http.Get(logs.HistoricURLs[0])
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		io.Copy(c.Out, resp.Body)
+	} else {
+		warn("No logs found for app component")
+	}
+
+	return nil
+}
+
+func parseAppSpec(spec []byte) (*godo.AppSpec, error) {
```

Comment:
Just a suggestion:

Might be nice to do some additional validation here with some feedback for the user. This will send and valid yaml or json as `{"spec":{"name":""}}` to the API.

I could even imagine spec validate sub-command being useful.

Reply:
The validation currently happens on the server side, so errors will actually flow though. One the reasons we don't want to have client side validation is to allow that spec to change while letting the backend handle all the backwards compatibility. Though there will need to be some sort of coordination between the backend and the generated spec types in godo.

The spec validate subcommand is a good idea and something I can follow up once we add support for an explicit validation endpoint on the backend.